src/conn.c

```diff
@@ -100,16 +100,23 @@ static void gateway_handle_cb(struct handle *req,
 	conn__stop(c);
 }
 
-static void closeCb(struct transport *transport)
+static void conn_defer_close_cb(uv_handle_t *t)
```

Comment:
It is a pity that we have to create so many callbacks just to move data around and to call the functions with the appropriate arguments. This is mostly unavoidable because you cannot inline functions in C, but there are also some cases that we could simplify. For example, we could change `close_cb` to get a `uv_handle_t` if it makes sense. That we we can remove one layer of callbacks. Probably there are other opportunities like this but I am not sure if those involve a larger refactor.

Reply:
I agree it's a pity to have so many little adapter callbacks but I feel like it's the least bad option in this case. I'm reluctant to change the signature of the conn close callback because it sort of leaks implementation details about the closing sequence into the interface of this module. You shorten conn.c by one function, but then all callsites of conn__start have to do the work of retrieving the conn from the handle, and if you want to know why the handle is there you have to go read conn.c.